packages/pglite-sync/src/index.ts

```diff
@@ -113,123 +156,183 @@ async function createPlugin(
+
+          if (truncateNeeded) {
+            truncateNeeded = false
+            // TODO: sync into shadow table and reference count
+            // for now just clear the whole table - will break
+            // cases with multiple shapes on the same table
+            await tx.exec(`DELETE FROM ${options.table};`)
+            if (options.shapeKey) {
+              await deleteShapeSubscriptionState({
+                pg: tx,
+                metadataSchema,
+                shapeKey: options.shapeKey,
+              })
+            }
+          }
+
+          if (doCopy) {
+            // We can do a `COPY FROM` to insert the initial data
+            // Split messageAggregator into initial inserts and remaining messages
+            const initialInserts: InsertChangeMessage[] = []
+            const remainingMessages: ChangeMessage<any>[] = []
+            let foundNonInsert = false
+            for (const message of messageAggregator) {
+              if (!foundNonInsert && message.headers.operation === 'insert') {
+                initialInserts.push(message as InsertChangeMessage)
+              } else {
+                foundNonInsert = true
+                remainingMessages.push(message)
+              }
+            }
+            if (initialInserts.length > 0) {
+              // As `COPY FROM` doesn't trigger a NOTIFY, we pop
+              // the last insert message and and add it to the be beginning
+              // of the remaining messages to be applied after the `COPY FROM`
+              remainingMessages.unshift(initialInserts.pop()!)
+            }
+            messageAggregator = remainingMessages
+
+            // Do the `COPY FROM` with initial inserts
+            if (initialInserts.length > 0) {
+              applyMessagesToTableWithCopy({
+                pg: tx,
+                table: options.table,
+                schema: options.schema,
+                messages: initialInserts as InsertChangeMessage[],
+                mapColumns: options.mapColumns,
+                primaryKey: options.primaryKey,
+                debug,
+              })
+              // We don't want to do a `COPY FROM` again after that
+              doCopy = false
+            }
+          }
+
+          for (const changeMessage of messageAggregator) {
+            await applyMessageToTable({
+              pg: tx,
+              table: options.table,
+              schema: options.schema,
+              message: changeMessage,
+              mapColumns: options.mapColumns,
+              primaryKey: options.primaryKey,
+              debug,
+            })
+          }
+
+          if (
+            options.shapeKey &&
+            messageAggregator.length > 0 &&
+            stream.shapeHandle !== undefined
+          ) {
+            await updateShapeSubscriptionState({
+              pg: tx,
+              metadataSchema,
+              shapeKey: options.shapeKey,
+              shapeId: stream.shapeHandle,
```

Comment:
I think this (`shapeHandle`) should be stored in a const at the start of the `commit` call for good measure - to avoid any changes to it in between applying messages and running async stuff.